### DISCUSSION TITLE
Question on using formula fields that return dates in Calc Item Where Clause

### DISCUSSION THREAD

**Discussion by jkenzer** [jamessimone/apex-rollup#120] [Q&A] [answered]
I'm trying to determine why I'm not getting the results I'm expecting and I wonder if it's because I'm using a formula field on the child record that returns a date. I know formulas aren't evaluated unless they are needed, like when displaying on a page or in a report. So I'm wondering if that is causing the issue?

Here's the scenario:

I'm trying to roll up fund grant amounts (child object) to funds (parent object) for a given time period. On the child object, I'm using Calc Item Where Clause = Grant_Date_From_Grant__c >= 2021-04-01. Grant_Date_From_Grant__c is a formula field that looks at a parent object, grants, to get the date. The fund grants object is a junction object between Grants and Funds.

The roll-up is working, but it's getting unexpected amounts. For example, I have one grant for $35,000 but I keep getting a value rolled up of $36,500. I've re-created the query in the developer console and in a report and get $35,000.

Here are my settings for this rollup:

![image](https://user-images.githubusercontent.com/643517/122088459-9574c100-cdba-11eb-9f3d-aec09669820d.png)

And here is a quick ERD:

![FGAs](https://user-images.githubusercontent.com/643517/122089406-94905f00-cdbb-11eb-92c9-3c7946f6bcaa.jpg)

**Comment by jamessimone** [comment, marked as answer]
@jkenzer have you cleared out the rollup field and recalculated using the rollup app? We have several other NPSP rollup orgs using formula fields with dates in the `Calc Item Where Clause`, so that shouldn't be an issue. Typically, for updates to the child objects (`Fund Grant Association` here), only when the `Grant Amount` field has been updated will that value get incremented to the existing parent-level value (which is why the full recalc using the app is a pre-req).

Let me know if the full recalculation is still yielding some bizarre value, like $36,500! If that ends up being the case, modifying the `Org_Defaults` rollup control Custom Metadata Type record to enable `Is Rollup Logging Enabled` may help uncover some additional insight.

Thanks!

**Reply by jkenzer** [reply]
@jamessimone I've been clearing and using the LWC button on the parent (Funds) to recalculate. Should I try from the Rollup app instead?

**Reply by jamessimone** [reply]
@jkenzer the button is a relatively new addition. I would be curious to know if there's any deviation between the app and the button. Off the top of my head, it's _not_ obvious why there would be a difference, because the button _should_ work via the same underlying code, but it's a good experiment. Otherwise, I'd be curious to hear a bit more about what's in the logs - it could be there's a bug in the way that the Calc Item Where Clause is being applied.

**Reply by jkenzer** [reply]
@jamessimone Interesting. When I ran it from the Rollup app, I did get the correct value. I'll reset it and run it both ways a few times to verify that the results are different between the two methods. I was under the impression that running it from the parent using the LWC would just re-run it for the parent, but it looks like both methods re-run it for all records?

I enabled logging, but I'm not sure where to find any useful information. I'm reviewing the normal apex log messages via the developer console. Since running it creates 69 batches, I get a new log file every three seconds. None of them appear to have Debug messages so I'm wondering if I'm looking in the wrong spot?

**Reply by jamessimone** [reply]
Which component are you dropping onto the parent flexipage? It should just be the button that says “recalculate rollups” - let me know if that’s the one you’re using and you’re still seeing it fully recalc for all records.

you are looking in the right place for those logs … there should definitely be debug statements that start “Rollup: “ with info about what stage in the rollup you’re at for each parent record and for each configured rollup. Do you have the Rollup Settings custom setting org default enabled? Just want to be sure since that could also be affecting things.

**Reply by jkenzer** [reply]
@jamessimone I'm embarrassed that I'm having such a hard time getting this to work. I guess I should probably start digging into the code to see if I can trace where things are going wrong. It's entirely possible that I have some code or automation that is causing conflicts. My org isn't that old, but it is highly customized, especially related to these three objects. I also have NPSP installed which has a lot of bloat when it comes to code running in the background.

On the LWC, I only see one button available. Here's a screenshot:

![image](https://user-images.githubusercontent.com/643517/122146870-a4845f00-ce0c-11eb-9563-8a2190520167.png)

This is this component, right? 

`/main/rollup/app/lwc/rollupForceRecalculation`

When I run it, I see 69 batches running in Apex jobs. This is the same amount of batches when I run it through the Rollup app.

![image](https://user-images.githubusercontent.com/643517/122147020-f3ca8f80-ce0c-11eb-8844-3a5718ba364a.png)

This particular test record has 33 Fund Grant Associations, so it should only be one batch.

I do have logging turned on via the Org Default. I will try toggling it off and back on and see if that helps.

![image](https://user-images.githubusercontent.com/643517/122147270-69cef680-ce0d-11eb-95ff-7a37a4ce13a3.png)

I super appreciate your responses. I'm currently using DLRS and have been hitting performance bottlenecks and would love to get some more performant rollups going.

**Reply by jkenzer** [reply]
@jamessimone Okay, I have logs with debugs from SerialBatchApexRangeChunkHandler. Looks like a JSON object for each record that it's pulling. Should it have records that are outside my date range? I see many that fall before my date range. This is the rollup I'm running (it's just a simple count from last fiscal year and I know this field, Number of Grants Last Year, isn't being affected by any other process because I just created it):

![image](https://user-images.githubusercontent.com/643517/122150219-275be880-ce12-11eb-9c6e-0096e58d5846.png)

And I'm seeing records like:

![image](https://user-images.githubusercontent.com/643517/122150289-40649980-ce12-11eb-9bcc-35acf50cb9aa.png)

This rollup is undercounting the number of records. Is it possible that it's silently failing on the process? The Apex jobs says 69 of 69 batches ran. There should be a total of 18,162 lines counted for that time range but only 3,823 are being rolled up.

**Reply by jamessimone** [reply]
@jkenzer please don’t apologize! Clearly something is amiss and I will have to track down the source of the issue. Let me try a few tests tomorrow morning — based on your screenshots and feedback this should be easy to track down. I have one other issue for another NPSP user that I’m hoping to track down at the same time — both will be part of the next release.

Don’t worry. I’m dedicated to ensuring everything works flawlessly for you!

**Reply by jamessimone** [reply]
@jkenzer I've created #123 & #123 to address the issues you're experiencing. I have them both fixed, locally, and plan to have those fixes (as well as several others) delivered as part of `v1.2.30` (to be released, latest, by this Friday).

**Reply by jamessimone** [reply]
@jkenzer both the issue with the recalc from parent button and the date filtering issue should be solved with the latest release. Please let me know!

**Reply by jkenzer** [reply]
Confirmed that both issues were fixed in V1.2.30.
